File: Codes/Multi Step Generator - 3D GMM/potential/GMM_3D.m
classdef GMM_3D
    % GMM_3D Potential Energy System Class (Auto-Geometric Bridge)
    %
    % Logic:
    %   U_lambda(x) = (1 - lambda) * U_base(x) + lambda * U_target(x)
    %   Lambda Schedule: 0 -> [Geometric Progression] -> 1
    %
    %   The number of steps M is calculated automatically such that the
    %   ratio between adjacent non-zero lambdas never exceeds BETA_RATE.
    
    properties
        % ---- 1. Configurable Parameters ----
        MU_SIZE     = 50000;
        MEAN        = 0;        
        SIGMA       = 2.5;      
        
        % ---- 2. Ladder Constraints ----
        % BETA_MIN: The starting value of the geometric section (Updated to 0.2)
        % BETA_RATE: The maximum allowed growth rate (ratio) between steps.
        BETA_MIN    = 0.08;      
        BETA_RATE   = 4;     
        
        % ---- 3. Derived Ladder Properties ----
        PT_M                    % Number of geometric intervals (Calculated)
        BETA_LIST               % [PT_M + 1, 1] Full lambda sequence including 0
        
        % ---- 4. Dynamics Parameters ----
        DT          = 5e-3;     
        COR_TIME    = 2;        
        CONST_SWAP  = 0.05;      
        WARM_TIME   = 200;      
        
        % ---- 5. Internal ----
        SWAP_INTERVAL_STEPS     
        TARGET_MEANS            
        TARGET_PRECISIONS       
        TARGET_LOG_DETS         
        TARGET_LOG_WEIGHTS      
    end
    
    properties (Constant)
        NAME = '3D Anisotropic GMM (Auto-Geometric)';
        ABBR = 'GM';
        DIM  = 3;
        
        X_LIM_COMPUTE = [-12, 12];
        Y_LIM_COMPUTE = [-12, 12];
        Z_LIM_COMPUTE = [-12, 12];
        
        X_LIM_PLOT    = [-8, 8];
        Y_LIM_PLOT    = [-8, 8];
        Z_LIM_PLOT    = [-8, 8];
    end
    
    methods
        function obj = GMM_3D(varargin)
            p = inputParser;
            addParameter(p, 'MU_SIZE', obj.MU_SIZE);
            addParameter(p, 'MEAN', obj.MEAN);
            addParameter(p, 'SIGMA', obj.SIGMA);
            addParameter(p, 'BETA_MIN', obj.BETA_MIN);
            addParameter(p, 'BETA_RATE', obj.BETA_RATE);
            addParameter(p, 'DT', obj.DT);
            addParameter(p, 'COR_TIME', obj.COR_TIME);
            addParameter(p, 'CONST_SWAP', obj.CONST_SWAP);
            addParameter(p, 'WARM_TIME', obj.WARM_TIME);
            
            parse(p, varargin{:});
            fields = fieldnames(p.Results);
            for i = 1:length(fields), obj.(fields{i}) = p.Results.(fields{i}); end
            
            % ---- 1. Calculate Minimum M ----
            % We need a sequence start=BETA_MIN, end=1.0 with ratio <= BETA_RATE.
            % M >= 1 + log(1/start) / log(rate)
            
            numer = log(1.0 / obj.BETA_MIN);
            denom = log(obj.BETA_RATE);
            min_steps = ceil(1 + numer / denom);
            
            obj.PT_M = max(2, min_steps); % Ensure at least start and end
            
            % ---- 2. Construct Geometric Ladder ----
            % logspace(a, b, n) generates n points between 10^a and 10^b
            geom_part = logspace(log10(obj.BETA_MIN), log10(1.0), obj.PT_M)';
            
            % ---- 3. Full Ladder (Prepend 0) ----
            obj.BETA_LIST = [0; geom_part];
            
            % Dynamics
            swap_time = obj.CONST_SWAP * obj.COR_TIME;
            obj.SWAP_INTERVAL_STEPS = max(1, round(swap_time / obj.DT));
            
            obj = obj.init_target_parameters();
        end
        
        function obj = init_target_parameters(obj)
            % (Standard GMM Initialization - Unchanged)
            K = 6; L = 6.0;
            obj.TARGET_MEANS = [L,0,0; -L,0,0; 0,L,0; 0,-L,0; 0,0,L; 0,0,-L];
            obj.TARGET_PRECISIONS = zeros(K, 3, 3);
            obj.TARGET_LOG_DETS   = zeros(K, 1);
            
            shapes = [0.3,0.3,0.8; 0.3,0.8,0.3; 0.8,0.3,0.3; 0.3,0.6,0.6; 0.4,0.4,0.4; 0.25,0.8,0.25];
            angles = [0,0,0; 45,45,0; 30,0,30; 90,45,0; 15,15,15; 60,0,-60];
            
            for k = 1:K
                lambda_val = diag(shapes(k, :).^2);
                phi=deg2rad(angles(k,1)); theta=deg2rad(angles(k,2)); psi=deg2rad(angles(k,3));
                Rz = [cos(phi) -sin(phi) 0; sin(phi) cos(phi) 0; 0 0 1];
                Ry = [cos(theta) 0 sin(theta); 0 1 0; -sin(theta) 0 cos(theta)];
                Rx = [1 0 0; 0 cos(psi) -sin(psi); 0 sin(psi) cos(psi)];
                R = Rz * Ry * Rx;
                Sigma = R * lambda_val * R';
                obj.TARGET_PRECISIONS(k, :, :) = inv(Sigma);
                obj.TARGET_LOG_DETS(k) = log(det(Sigma));
            end
            weights = ones(K, 1) / K;
            obj.TARGET_LOG_WEIGHTS = log(weights);
        end
        
        function u = compute_potential(obj, x)
            u = obj.compute_potential_mixed(x, 1.0);
        end
        
        function u_mix = compute_potential_mixed(obj, x, lambda)
            % U_lam = (1-lam)U0 + lam*U1
            z = (x - obj.MEAN) ./ obj.SIGMA;
            u_base = 0.5 * sum(z.^2, 2);
            
            M_samples = size(x, 1); K = 6;
            log_probs = zeros(M_samples, K);
            const_term = -0.5 * obj.DIM * log(2 * pi);
            
            for k = 1:K
                diff = x - obj.TARGET_MEANS(k, :);
                P = squeeze(obj.TARGET_PRECISIONS(k, :, :));
                term = diff * P; 
                mahalanobis = -0.5 * sum(term .* diff, 2);
                log_gauss = const_term - 0.5 * obj.TARGET_LOG_DETS(k) + mahalanobis;
                log_probs(:, k) = obj.TARGET_LOG_WEIGHTS(k) + log_gauss;
            end
            max_log = max(log_probs, [], 2);
            log_p = max_log + log(sum(exp(log_probs - max_log), 2));
            u_target = -log_p;
            
            u_mix = (1.0 - lambda) .* u_base + lambda .* u_target;
        end
        
        function grad_mix = compute_gradient_mixed(obj, x, lambda)
            grad_base = (x - obj.MEAN) ./ (obj.SIGMA^2);
            
            M_samples = size(x, 1); K = 6;
            log_joint = zeros(M_samples, K);
            const_term = -0.5 * obj.DIM * log(2 * pi);
            
            for k = 1:K
                diff = x - obj.TARGET_MEANS(k, :);
                P = squeeze(obj.TARGET_PRECISIONS(k, :, :));
                mahalanobis = -0.5 * sum((diff * P) .* diff, 2);
                log_joint(:, k) = obj.TARGET_LOG_WEIGHTS(k) + const_term - 0.5 * obj.TARGET_LOG_DETS(k) + mahalanobis;
            end
            max_log = max(log_joint, [], 2);
            log_sum = max_log + log(sum(exp(log_joint - max_log), 2));
            responsibilities = exp(log_joint - log_sum); 
            
            grad_target = zeros(M_samples, 3);
            for k = 1:K
                diff = x - obj.TARGET_MEANS(k, :);
                P = squeeze(obj.TARGET_PRECISIONS(k, :, :));
                grad_comp = diff * P;
                grad_target = grad_target + responsibilities(:, k) .* grad_comp;
            end
            
            grad_mix = (1.0 - lambda) .* grad_base + lambda .* grad_target;
        end
        
        function x_next = integrator(obj, x)
            lambdas = obj.BETA_LIST; 
            dt      = obj.DT;
            grad = obj.compute_gradient_mixed(x, lambdas);
            
            raw_drift = -grad * dt;
            taming_factor = 1 + dt * abs(grad);
            tamed_drift = raw_drift ./ taming_factor;
            
            noise = randn(size(x));
            sigma = sqrt(2 * dt); 
            diffusion = sigma .* noise;
            
            x_next = x + tamed_drift + diffusion;
        end
        
        function grad = compute_gradient(obj, x)
            grad = obj.compute_gradient_mixed(x, 1.0);
        end
        
        function disp_info(obj)
            fprintf('--- %s (%s) ---\n', obj.NAME, obj.ABBR);
            fprintf('Strategy: Auto-Geometric Bridge (Rate <= %.2f)\n', obj.BETA_RATE);
            fprintf('Params: Min=%.3f, Max=1.0\n', obj.BETA_MIN);
            fprintf('Ladder Steps (M): %d (Geometric) -> Total Replicas: %d\n', obj.PT_M, length(obj.BETA_LIST));
            
            % [UPDATED]: Print Full Ladder
            fprintf('Lambda Sequence (Full): %s\n', mat2str(obj.BETA_LIST', 4));
            
            fprintf('Base Sigma: %.2f\n', obj.SIGMA);
            fprintf('----------------------------------\n');
        end
    end
end

File: Codes/Multi Step Generator - 3D GMM/potential/GMM_3D.py
import torch
import math


# =========================================================================
# 1. JIT Compiled Calculation Kernel (Potential)
# =========================================================================

@torch.jit.script
def _gmm_3d_jit(x: torch.Tensor,
                means: torch.Tensor,
                precs: torch.Tensor,
                log_dets: torch.Tensor,
                log_weights: torch.Tensor) -> torch.Tensor:
    """
    JIT-compiled kernel for 3D Anisotropic GMM Potential.
    Input: x [N, 3]
    Output: V [N] (Potential energy)
    """
    N = x.shape[0]
    K = means.shape[0]

    # Constants: log(2*pi) * 3 / 2
    const_term = -0.5 * 3.0 * math.log(2.0 * math.pi)

    # Store log probabilities for each component [N, K]
    log_probs = torch.zeros((N, K), dtype=x.dtype, device=x.device)

    for k in range(K):
        # 1. Diff: (x - mu_k) -> [N, 3]
        diff = x - means[k]

        # 2. Mahalanobis Term: -0.5 * (x-mu)^T * P * (x-mu)
        P = precs[k]  # [3, 3]
        term = torch.mm(diff, P)  # [N, 3]
        mahalanobis = -0.5 * torch.sum(term * diff, dim=1)

        # 3. Log Gaussian PDF
        log_gauss = const_term - 0.5 * log_dets[k] + mahalanobis

        # 4. Log Joint: log(w_k) + log N
        log_probs[:, k] = log_weights[k] + log_gauss

    # LogSumExp across components -> [N]
    log_p = torch.logsumexp(log_probs, dim=1)

    # U(x) = -log p(x)
    return -log_p


# =========================================================================
# 2. JIT Compiled Calculation Kernel (Gradient)
# =========================================================================

@torch.jit.script
def _gmm_3d_grad_jit(x: torch.Tensor,
                     means: torch.Tensor,
                     precs: torch.Tensor,
                     log_dets: torch.Tensor,
                     log_weights: torch.Tensor) -> torch.Tensor:
    """
    JIT-compiled kernel for Gradient of 3D Anisotropic GMM Potential.
    """
    N = x.shape[0]
    K = means.shape[0]
    const_term = -0.5 * 3.0 * math.log(2.0 * math.pi)

    # 1. Compute Log Probabilities
    log_probs = torch.zeros((N, K), dtype=x.dtype, device=x.device)

    for k in range(K):
        diff = x - means[k]
        P = precs[k]
        term = torch.mm(diff, P)
        mahalanobis = -0.5 * torch.sum(term * diff, dim=1)
        log_gauss = const_term - 0.5 * log_dets[k] + mahalanobis
        log_probs[:, k] = log_weights[k] + log_gauss

    # 2. Compute Responsibilities
    log_p = torch.logsumexp(log_probs, dim=1, keepdim=True)
    responsibilities = torch.exp(log_probs - log_p)

    # 3. Compute Weighted Gradient
    grad = torch.zeros_like(x)

    for k in range(K):
        diff = x - means[k]
        P = precs[k]
        grad_comp = torch.mm(diff, P)
        w = responsibilities[:, k].unsqueeze(1)
        grad += w * grad_comp

    return grad


# =========================================================================
# 3. Potential Energy Class
# =========================================================================

class GMM_3D:
    """
    3D ANISOTROPIC GMM Potential Energy System.
    Note: Beta Ladder configuration is now handled externally (parameters.py).
    """

    NAME = '3D Anisotropic GMM'
    ABBR = 'GM'
    DIM = 3

    # Base Distribution Parameters
    BASE_MEAN = 0.0
    BASE_SIGMA = 2.5

    # Rejuvenation Parameters (Default for this potential)
    DT = 5e-3
    N_ITER = 400

    # Computing/Plotting Boundaries
    X_LIM_PLOT = [-8, 8]
    Y_LIM_PLOT = [-8, 8]
    Z_LIM_PLOT = [-8, 8]
    X_LIM_COMPUTE = [-12, 12]
    Y_LIM_COMPUTE = [-12, 12]
    Z_LIM_COMPUTE = [-12, 12]

    def __init__(self, device: str = 'cpu'):
        self.device = device

        # Initialize Gaussian components
        self._init_target_parameters()

    def to(self, device):
        """
        Moves all internal tensors to the specified device.
        Mimics the behavior of nn.Module.to(device).
        """
        self.device = device

        # Move GMM Parameters
        if hasattr(self, 'means'):
            self.means = self.means.to(device)
            self.precs = self.precs.to(device)
            self.log_dets = self.log_dets.to(device)
            self.log_weights = self.log_weights.to(device)

        return self

    def _init_target_parameters(self):
        """
        Constructs 6 anisotropic Gaussian components in an octahedral layout.
        """
        K = 6
        L = 6.0

        means_list = [
            [L, 0, 0], [-L, 0, 0], [0, L, 0], [0, -L, 0], [0, 0, L], [0, 0, -L]
        ]
        shapes = [
            [0.3, 0.3, 0.8], [0.3, 0.8, 0.3], [0.8, 0.3, 0.3],
            [0.3, 0.6, 0.6], [0.4, 0.4, 0.4], [0.25, 0.8, 0.25]
        ]
        angles_deg = [
            [0, 0, 0], [45, 45, 0], [30, 0, 30],
            [90, 45, 0], [15, 15, 15], [60, 0, -60]
        ]

        precs_list = []
        log_dets_list = []

        for k in range(K):
            lam_vals = [s ** 2 for s in shapes[k]]
            Lambda = torch.diag(torch.tensor(lam_vals, dtype=torch.float32))

            phi, theta, psi = [math.radians(a) for a in angles_deg[k]]

            Rz = torch.tensor([[math.cos(phi), -math.sin(phi), 0],
                               [math.sin(phi), math.cos(phi), 0],
                               [0, 0, 1]], dtype=torch.float32)
            Ry = torch.tensor([[math.cos(theta), 0, math.sin(theta)],
                               [0, 1, 0],
                               [-math.sin(theta), 0, math.cos(theta)]], dtype=torch.float32)
            Rx = torch.tensor([[1, 0, 0],
                               [0, math.cos(psi), -math.sin(psi)],
                               [0, math.sin(psi), math.cos(psi)]], dtype=torch.float32)

            R = torch.mm(Rz, torch.mm(Ry, Rx))
            Sigma = torch.mm(R, torch.mm(Lambda, R.t()))
            Prec = torch.inverse(Sigma)
            log_det = torch.log(torch.det(Sigma))

            precs_list.append(Prec)
            log_dets_list.append(log_det)

        # Move to initial device
        self.means = torch.tensor(means_list, dtype=torch.float32).to(self.device)
        self.precs = torch.stack(precs_list).to(self.device)
        self.log_dets = torch.stack(log_dets_list).to(self.device)

        weights = torch.ones(K, dtype=torch.float32) / K
        self.log_weights = torch.log(weights).to(self.device)

    # ---- Potential Functions ----

    def compute_potential_base(self, x: torch.Tensor) -> torch.Tensor:
        """
        U_base: Gaussian Potential
        U(x) = 0.5 * ||(x - mu)/sigma||^2
        """
        if x.device != torch.device(self.device): x = x.to(self.device)
        z = (x - self.BASE_MEAN) / self.BASE_SIGMA
        return 0.5 * torch.sum(z ** 2, dim=1)

    def compute_potential_target(self, x: torch.Tensor) -> torch.Tensor:
        """
        U_target: Pure GMM Potential
        """
        if x.device != torch.device(self.device): x = x.to(self.device)
        return _gmm_3d_jit(x, self.means, self.precs, self.log_dets, self.log_weights)

    def compute_potential_mixed(self, x: torch.Tensor, lam: float) -> torch.Tensor:
        """
        U_mix = (1-lam)*U_base + lam*U_target
        Used for linear interpolation training.
        """
        u0 = self.compute_potential_base(x)
        u1 = self.compute_potential_target(x)
        return (1.0 - lam) * u0 + lam * u1

    # ---- Gradient Functions (for MALA) ----

    def compute_gradient_base(self, x: torch.Tensor) -> torch.Tensor:
        """
        grad U_base(x) = (x - mu) / sigma^2
        """
        if x.device != torch.device(self.device): x = x.to(self.device)
        return (x - self.BASE_MEAN) / (self.BASE_SIGMA ** 2)

    def compute_gradient_target(self, x: torch.Tensor) -> torch.Tensor:
        """
        grad U_target(x) using JIT kernel
        """
        if x.device != torch.device(self.device): x = x.to(self.device)
        return _gmm_3d_grad_jit(x, self.means, self.precs, self.log_dets, self.log_weights)

    def compute_gradient_mixed(self, x: torch.Tensor, lam: float) -> torch.Tensor:
        """
        grad U_mix = (1-lam)*grad_U_base + lam*grad_U_target
        """
        g0 = self.compute_gradient_base(x)
        g1 = self.compute_gradient_target(x)
        return (1.0 - lam) * g0 + lam * g1